corehq/form_processor/submission_post.py

```diff
@@ -500,14 +497,30 @@ def index_case_search(instance, case_models):
 
     @staticmethod
     @tracer.wrap(name='submission.process_cases_and_stock')
-    def process_xforms_for_cases(xforms, case_db, timing_context=None):
+    def process_xforms_for_cases(xforms, case_db, domain=None, timing_context=None):
         from casexml.apps.case.xform import process_cases_with_casedb
         from corehq.apps.commtrack.processing import process_stock
 
-        timing_context = timing_context or TimingContext()
-
         instance = xforms[0]
 
+        # INSERT "NOT" HERE when done w testing
+        if ALLOW_SUBMISSION_WITHOUT_METADATA.enabled(domain):
+            form_data = instance.form_data
+            if 'case' in form_data:
+                if 'create' in form_data['case']:
+                    for tag in required_create_tags:
+                        if tag not in form_data['case']['create']:
+                            raise MetadataError('One or more required tags missing in the create node')
+                elif 'update' in form_data['case'] and not isinstance(form_data['case']['update'], dict):
+                    raise MetadataError('Missing nodes in the update node.')
+            if instance.metadata:
+                for attr in required_metadata_tags:
+                    if getattr(instance.metadata, attr) is None:
+                        raise MetadataError('One or more required tags missing in the metadata node')
+            else:
+                raise MetadataError('Missing metadata node.')
```

Comment:
At least this (potentially the whole validation block) should be moved out of the case processing function and placed somewhere before we call get_locked_forms so that we can avoid doing the extra work involved in getting locks etc.